### PR TITLE
[5.1] Throw exception if paginator exceeds its maximum amount of pages

### LINKED ISSUE #8252: Paginator Exceeding Pages
If you have 20 items per page on the paginator and you have 21 items, if someone else comes along and deletes one item leaving only 20, now there's only 1 page if the other person then refreshes they are still on page 2 and the paginator returns no results.

```php
$model->paginate(20);
```

Should the paginator not throw an Exception to catch them being out of bounds, or as most paginators do show the last possible page if there are any results at all.

### PR DESCRIPTION
**Pull request by Sladewill**
Push to 5.1 instead of #9302. Fixes #8252.

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -267,9 +267,17 @@ public function paginate($perPage = null, $columns = ['*'], $pageName = 'page')
...
+        $lastPage = ceil($total / $perPage);
+
+        if ($page > $lastPage) {
+            throw new PaginatorLastPageExceededException;
+        }
+
         $this->query->forPage(
-            $page = Paginator::resolveCurrentPage($pageName),
-            $perPage = $perPage ?: $this->model->getPerPage()
+            $page,
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: Please just one-line this function call.

File: src/Illuminate/Database/Eloquent/PaginatorLastPageExceededException.php [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?php namespace Illuminate\Database\Eloquent;
+
+use RuntimeException;
+
+class PaginatorLastPageExceededException extends RuntimeException
+{
+
+}
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/PaginatorLastPageExceededException.php]: cs

**Comment by Sladewill** [comment]
Should be fixed now.

File: src/Illuminate/Database/Eloquent/PaginatorLastPageExceededException.php [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<?php namespace Illuminate\Database\Eloquent;
+
+use RuntimeException;
+
+class PaginatorLastPageExceededException extends RuntimeException
+{
+
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/PaginatorLastPageExceededException.php]: no blank line here please

**Comment by Sladewill** [comment]
That should be fixed now as well.

**Comment by taylorotwell** [comment]
Going to hold off on this for now as existing application's won't be expecting an exception. It seems like we just send them to the last possible page.

**Comment by GrahamCampbell** [comment]
@taylorotwell How about getting this in on 5.2?

**Comment by Sladewill** [comment]
Tbh it could just redirect, then move to throwing exception afterwards, but either way this is an annoying issue to have to get around using hacky ways.

**Comment by GrahamCampbell** [comment]
> Tbh it could just redirect, then move to throwing exception afterwards,

How the application behaves is up to you. The code in there doesn't change regarding an exception.

**Comment by cviebrock** [comment]
My 2 cents: if you request page 3 of a 2-page result set, then I feel the Paginator instance should keep the items as an empty array (as it does now), but should return null if you ask for the current page (and null for the next or previous pages as well, perhaps).

**Pull request closed without merging**